### PR TITLE
Add Scope-Bounded Capabilities proposal

### PR DESCRIPTION
**Pull request by anandabits**

### CODE REVIEW

**Comment by anandabits** [comment]
@BasThomas the zip contains a playground that the proposal links to.

File: proposals/NNNN-scope-bounded-capabilities.md [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+
+With one exception, this proposal will not cause any Swift 3 source to fail to compile but will produce different behavior in four cases.  In all cases a mechanical migration is possible.
+
+### `fileprivate`
+
+This proposal deprecates `fileprivate`.  This keyword should be supported for at least one release with a deprecation warning and eventually removed.
+
+### `private`
+
+This proposal reverts the meaning of `private` to file scope.  This will not break any code but many users will want to convert existing uses of `private` to `scoped` during migration to Swift 4.
```
Comment (JustinJiaDev) [inline proposals/NNNN-scope-bounded-capabilities.md]:
```
class Foo {
    private bar() { ... }
}

extension Foo {
    private bar() { ... }
}
```

Will break.

**Comment by anandabits** [comment]
Thanks for pointing that out @JustinJiaDev.  I have updated the proposal to clarify this.

**Comment by CodaFi** [comment]
With @JustinJiaDev's comments, I think you should go back and consider the full source-breakage that will result from this proposal.  Now that it is not additive, the stakes for its inclusion go up significantly.

**Comment by anandabits** [comment]
@CodaFi I copied the latest template which includes a "Source compatibility" section.  Isn't that the same as "Impact on Existing Code"?

All drafts of this proposal have explicitly acknowledged the source breakage that will occur, most of which is a change in semantics rather than fail-to-compile.  All breakages will not be affected in Swift 3.1 compatibility mode and are easy to mechanically migrate.  

I also believe that all of the things this proposal breaks are things that we *want* to break eventually regardless of what the final state looks like.  Swift 4 is the right time to have this discussion.  If it were merely an additive change Swift 4 would not be the right time to have this discussion.

If you believe there are cases of source breakage which are not explicitly covered please at least point in the right direction.

**Comment by anandabits** [comment]
Out of scope for Swift 4.

**Pull request closed without merging**